From 42e34becc956db9da75c97b1c4701d4b58673f8a Mon Sep 17 00:00:00 2001
From: saranya <saranya.gopal@intel.com>
Date: Wed, 29 May 2019 11:30:51 +0530
Subject: [PATCH] roles: Fix issue with PM suspend

Role should be switched to host mode to
allow system suspend.

Signed-off-by: saranya <saranya.gopal@intel.com>
---
 .../usb/roles/intel-xhci-usb-role-switch.c    | 25 +++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/drivers/usb/roles/intel-xhci-usb-role-switch.c b/drivers/usb/roles/intel-xhci-usb-role-switch.c
index 8c3e9f8511c4..b22dafffcefe 100644
--- a/drivers/usb/roles/intel-xhci-usb-role-switch.c
+++ b/drivers/usb/roles/intel-xhci-usb-role-switch.c
@@ -195,6 +195,26 @@ static int intel_xhci_usb_remove(struct platform_device *pdev)
 	return 0;
 }
 
+#ifdef CONFIG_PM
+static int intel_xhci_usb_suspend(struct device *dev)
+{
+	pr_info("saranya:%s", __func__);
+        intel_xhci_usb_set_role(dev, USB_ROLE_HOST);
+        return 0;
+}
+
+static int intel_xhci_usb_resume(struct device *dev)
+{
+	pr_info("saranya:%s", __func__);
+        intel_xhci_usb_set_role(dev, USB_ROLE_DEVICE);
+        return 0;
+}
+
+static const struct dev_pm_ops intel_xhci_usb_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(intel_xhci_usb_suspend, intel_xhci_usb_resume)
+};
+#endif
+
 static const struct platform_device_id intel_xhci_usb_table[] = {
 	{ .name = DRV_NAME },
 	{}
@@ -208,6 +228,11 @@ static struct platform_driver intel_xhci_usb_driver = {
 	.id_table = intel_xhci_usb_table,
 	.probe = intel_xhci_usb_probe,
 	.remove = intel_xhci_usb_remove,
+#ifdef CONFIG_PM
+	.driver = {
+		.pm = &intel_xhci_usb_pm_ops,
+	},
+#endif
 };
 
 module_platform_driver(intel_xhci_usb_driver);
-- 
2.21.0

